### PR TITLE
Save requests to Redis

### PR DESCRIPTION
**Pull request by yoomlam**
## What was the problem?

For error recovery scenarios, we cannot "replay" request submitted to VRO's API b/c we don't currently have approval to persist PII data.

Associated tickets or Slack threads:
- #563

## How does this fix it?[^1]

Log the payload of incoming requests to in-memory (non-persisted) Redis so that we can manually resubmit those request when needed.

## How to test this PR
- `export COMPOSE_PROFILES=debug` -- the console container only starts if the docker-compose `debug` profile is enabled.
- Start VRO: `./gradlew app:dockerComposeUp`. Note the `vro-console-1` container (optionally check out the logs: `docker logs vro-console-1`).
- Using Swagger, test the 2 VRO endpoints
- Attach to the container: `docker attach vro-console-1` (Alternative: use `redis-cli` to run equivalent commands.)
- On the `groovy:000>` prompt, try the following:
```groovy
// List all the Redis keys available
groovy:000> redis.keys "*"

// "1234" is the claim id
groovy:000> redis.hgetall "tracking-1234"
// Shows the keys and values of the hash at Redis key "tracking-1234"

// Query the individual hash entries
groovy:000> redis.hget "tracking-1234", "submitted-claim"
groovy:000> redis.hget "tracking-1234", "submitted-pdf"
```

[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.

### CODE REVIEW

**Review by au2 — CHANGES_REQUESTED** [review]
Followed the instructions. After running the two end points the following I got. 1 and 3 are problematic.  Repeated with with a second try with 1235 and same results.

1) redis.hgetall "tracking-1234" "submitted-claim" returns null.  This was expected to be the incoming claim.
2) redis.hgetall "tracking-1234" "submitted-pdf" works as expected.
3) There is an additional "1234-pdf" key (groovy:000> redis.keys "*"
===> [1234-pdf, tracking-1234]). It has diagnosis and content fields. Which tells me it is the output of pdf generating microservice. I do not think we are expecting it, do we?

Save2Db replaces the body (which I want to fix at some point). Maybe that is the reason for 1.

**Comment by yoomlam** [comment]
> Followed the instructions. After running the two end points the following I got. 1 and 3 are problematic. Repeated with with a second try with 1235 and same results.
> 
> 1. redis.hgetall "tracking-1234" "submitted-claim" returns null.  This was expected to be the incoming claim.

Curious. I'll retest to see what I get.

> 3. There is an additional "1234-pdf" key (groovy:000> redis.keys "*"
>    ===> [1234-pdf, tracking-1234]). It has diagnosis and content fields. Which tells me it is the output of pdf generating microservice. I do not think we are expecting it, do we?

The `1234-pdf` hash is created by the pdf-generator microservice. This is expected.

> Save2Db replaces the body (which I want to fix at some point). Maybe that is the reason for 1.

**Comment by yoomlam** [comment]
@au2 I'm unable to reproduce, here's my output:
```
groovy:000> redis.keys "*"
===> []

// Submit to health-data-assessment in Swagger

groovy:000> redis.keys "*"
===> [tracking-1235]
groovy:000> redis.hgetall "tracking-1235"
===> [submitted-claim:{"recordId":null,"claimSubmissionId":"1235","idType":"va.gov-Form526Submission","incomingStatus":"submission","veteranIcn":"9000682","diagnosticCode":"7101","contentions":null}]
// Verified only 1 hash entry with key "submitted-claim"

// Submit same request to health-data-assessment in Swagger again

groovy:000> redis.hgetall "tracking-1235"
===> [submitted-claim:{"recordId":null,"claimSubmissionId":"1235","idType":"va.gov-Form526Submission","incomingStatus":"submission","veteranIcn":"9000682","diagnosticCode":"7101","contentions":null}]
// Verified no change to hash entry

// Submit to evidence-pdf in Swagger

groovy:000> redis.hgetall "tracking-1235"
===> [
submitted-pdf:{"claimSubmissionId":"1235","diagnosticCode":"7101","veteranInfo":{"first":"Joe","middle":"M","last":"Doe","suffix":"Jr","birthdate":"01/04/1971"},"evidence":{"medications":[{"status":"Active","notes":["string"],"description":"Hydrochlorothiazide 6.25 MG","refills":2,"asthmaRelevant":true,"duration":"string","authoredOn":"2021-04-05T23:00:00Z","dosageInstructions":["QID PRN","As directed by physician"],"route":"INHALATION ORAL"}],"conditions":[{"text":"string","code":"string","abatementDate":"string","status":"string","onsetDate":"string"}],"procedures":[{"text":"string","code":"string","status":"string","performedDate":"string","codeSystem":"string"}],"bp_readings":[{"date":"2020-08-05","diastolic":{"code":"8480-6","display":"Systolic blood pressure","unit":"mm[Hg]","value":110},"systolic":{"code":"8480-6","display":"Systolic blood pressure","unit":"mm[Hg]","value":110},"practitioner":"DR. THOMAS REYNOLDS PHD","organization":"WASHINGTON VA MEDICAL CENTER"}]},"status":null,"reason":null,"eventId":"1235"}, 
submitted-claim:{"recordId":null,"claimSubmissionId":"1235","idType":"va.gov-Form526Submission","incomingStatus":"submission","veteranIcn":"9000682","diagnosticCode":"7101","contentions":null}]
// Verified 2 hash entries with key "submitted-claim" and "submitted-pdf"

// Retrieve each hash entry individually:

groovy:000> redis.hget "tracking-1235", "submitted-claim"
===> {"recordId":null,"claimSubmissionId":"1235","idType":"va.gov-Form526Submission","incomingStatus":"submission","veteranIcn":"9000682","diagnosticCode":"7101","contentions":null}

groovy:000>  redis.hget "tracking-1235", "submitted-pdf"
===> {"claimSubmissionId":"1235","diagnosticCode":"7101","veteranInfo":{"first":"Joe","middle":"M","last":"Doe","suffix":"Jr","birthdate":"01/04/1971"},"evidence":{"medications":[{"status":"Active","notes":["string"],"description":"Hydrochlorothiazide 6.25 MG","refills":2,"asthmaRelevant":true,"duration":"string","authoredOn":"2021-04-05T23:00:00Z","dosageInstructions":["QID PRN","As directed by physician"],"route":"INHALATION ORAL"}],"conditions":[{"text":"string","code":"string","abatementDate":"string","status":"string","onsetDate":"string"}],"procedures":[{"text":"string","code":"string","status":"string","performedDate":"string","codeSystem":"string"}],"bp_readings":[{"date":"2020-08-05","diastolic":{"code":"8480-6","display":"Systolic blood pressure","unit":"mm[Hg]","value":110},"systolic":{"code":"8480-6","display":"Systolic blood pressure","unit":"mm[Hg]","value":110},"practitioner":"DR. THOMAS REYNOLDS PHD","organization":"WASHINGTON VA MEDICAL CENTER"}]},"status":null,"reason":null,"eventId":"1235"}
```

**Comment by yoomlam** [comment]
You can run the following to see what's being received by the 2 endpoints before you submit the requests in Swagger:
```
wireTap "claim-submitted"
wireTap "generate-pdf"
```

**Comment by au2** [comment]
@yoomlam  I see. You put the wiretap in configureRouteClaimSubmit. You need to put it in configureRouteClaimSubmitForFull and use full-health-data-assessment (that is the end point that is being used currently).  health-data-assessment is the original that do not use the filtering done by the python service.  I think we will remove the original one at some point.

BTW redis.hget "tracking-1235", "submitted-claim" appears to be returning output of save2db in your example. Which I expected I think. I really do not like save2db changing the body.

**Comment by yoomlam** [comment]
> @yoomlam I see. You put the wiretap in configureRouteClaimSubmit. You need to put it in configureRouteClaimSubmitForFull and use full-health-data-assessment (that is the end point that is being used currently). health-data-assessment is the original that do not use the filtering done by the python service. I think we will remove the original one at some point.

Added wireTap to `configureRouteClaimSubmitForFull` and retested with that endpoint.

> BTW redis.hget "tracking-1235", "submitted-claim" appears to be returning output of save2db in your example. Which I expected I think. I really do not like save2db changing the body.

Why does it appear like that?
Since the `recordId` is `null`, that means the message is being saved to Redis before it is sent to `saveToDbService`. After `saveToDbService`, the `recordId` is not null.
```
groovy:000> redis.hget "tracking-1235", "submitted-claim"
===> {"recordId":null,"claimSubmissionId":"1235","idType":"va.gov-Form526Submission","incomingStatus":"submission","veteranIcn":"9000682","diagnosticCode":"7101","contentions":null}
```

File: service/provider/src/main/java/gov/va/vro/service/provider/camel/PrimaryRoutes.java [hunk 1]
```diff
@@ -60,6 +64,7 @@ private void configureRouteClaimSubmitForFull() {
     // send JSON-string payload to RabbitMQ
     from(ENDPOINT_SUBMIT_CLAIM_FULL)
         .routeId("claim-submit-full")
+        .wireTap(wireTapTopicFor(INCOMING_CLAIM_WIRETAP))
         .process(FunctionProcessor.fromFunction(saveToDbService::insertClaim))
```
Comment (yoomlam) [inline service/provider/src/main/java/gov/va/vro/service/provider/camel/PrimaryRoutes.java:68]: wireTap before `saveToDbService`

**Review by au2 — APPROVED** [review]
works as expected.

**Pull request merged**